File: Easy/121.BestTimetoBuyandSellStock.js
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

var maxProfit = function (prices) {
  let [max, min] = [0, 0];
  for (let i = 0; i < prices.length; i++) {
	if (prices[i] < prices[min]) min = i;
	if (prices[i] - prices[min] > max) max = prices[i] - prices[min];
  }
  return max;
};

const testCases = [
  [[7, 1, 5, 3, 6, 4], 5],
  [[7, 6, 4, 3, 1], 0],
];

for (const test of testCases) {
  console.log(maxProfit(test[0]), "expected value = ", test[1]);
}

File: 121-Best-Time-to-Buy-and-Sell-Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0

        for n in prices:

            if n < min_price:
                min_price = n
            else:
                max_profit = max(max_profit, n - min_price)
        
        return max_profit